[AArch64][SME] Make getRegAllocationHints more specific for multi-vector loads

kmclaughlin-arm opened the pull request:
getRegAllocationHints looks for ZPR2StridedOrContiguous load instructions
which are used by FORM_TRANSPOSED_REG_TUPLE pseudos and adds all
strided registers from this class to the list of hints.
This patch changes getRegAllocationHints to restrict this list:
 - If the pseudo uses ZPRMul class, the first load must begin with a register
which is a multiple of 2 or 4.
 - Only add a hint if it is part of a sequence of registers that do not already
have any live intervals.

This also contains changes to suggest hints when the load instructions and
the FORM_TRANSPOSED pseudo use multi-vectors of different lengths,
e.g. a pseudo with a 4-vector sequence of registers formed of one column
extracted from four 2-vector loads.

kmclaughlin-arm committed:
Add uses of %scalable_arg in SME2 dot intrinsics tests

kmclaughlin-arm committed:
[AArch64][SME] Make getRegAllocationHints stricter for multi-vector loads.

getRegAllocationHints looks for ZPR2StridedOrContiguous load instructions
which are used by FORM_TRANSPOSED_REG_TUPLE pseudos and adds all
strided registers from this class to the list of hints.
This patch changes getRegAllocationHints to restrict this list:
 - If the pseudo uses ZPRMul class, the first load must begin with a register
   which is a multiple of 2 or 4.
 - Only add a strided register to the list if it does not already have any
   live intervals.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -1107,23 +1108,83 @@ bool AArch64RegisterInfo::getRegAllocationHints(
+        if (Matrix->isPhysRegUsed(Reg))
+          continue;
+
+        bool CanAssign = true;
+        for (unsigned Next = 1; Next < LdOps; ++Next) {
+          // Ensure we can assign enough registers from the list for all loads.
+          if (I + Next >= StridedOrder.size()) {
+            CanAssign = false;
+            break;
+          }
+          // Ensure the subsequent registers are not live and that the starting
+          // sub-registers are sequential.
+          MCPhysReg NextReg = StridedOrder[I + Next];
+          if (Matrix->isPhysRegUsed(NextReg) ||
+              (getSubReg(NextReg, AArch64::zsub0) != FirstReg + Next)) {
+            CanAssign = false;
+            break;
+          }
+        }
+        if (CanAssign)
+          Hints.push_back(Reg);
```
This loop can be written more compactly doing something along the lines of:
```
for (unsigned Next = 1; Next < LdOps; ++Next) {
  if (!is_contained(StridedOrder, Reg + Next))
    // cannot assign
}
```

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -1107,23 +1108,83 @@ bool AArch64RegisterInfo::getRegAllocationHints(
+    bool IsMulZPR = TupleID == AArch64::ZPR2Mul2RegClassID ||
+                    TupleID == AArch64::ZPR4Mul4RegClassID;
+
+    if (OpIdx == 1) {
```
This is making an assumption about the order in which the register allocator visits the operands and allocates registers. I think that instead you want to distinguish between the cases where:
* None of the operands for Use are allocated yet => then check if for each of the operands, we have consecutive tuple registers available, and if so then add the hint for the given OpIdx.
* Any of the operands for Use is already allocated => find the corresponding tuple reg for the current operand index, if available.

kmclaughlin-arm committed:
- Instead of finding a hint based on the operand index from the FORM_TRANSPOSED
  pseudo, find hints based on whether any operands have been allocated yet.
- Remove any_of which finds FORM_TRANSPOSED uses of VirtReg & instead iterate
  through all uses.
- Clean up sme2-multivec-regalloc.mir & add new test which changes the
  allocation order of the load instructions.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -1109,24 +1110,93 @@ bool AArch64RegisterInfo::getRegAllocationHints(
+        // Look for registers in StridedOrder which start with sub-registers
+        // following sequentially from FirstReg. If all are found and none are
+        // already live, add Reg to Hints.
+        MCPhysReg RegToAssign = Reg;
+        for (unsigned Next = 1; Next < LdOps; ++Next) {
+          MCPhysReg Strided = GetRegStartingAt(FirstReg + Next);
+          if (Strided == AArch64::NoRegister ||
+              Matrix->isPhysRegUsed(Strided)) {
+            RegToAssign = AArch64::NoRegister;
+            break;
+          }
+          if (Next == (unsigned)OpIdx - 1)
+            RegToAssign = Strided;
+        }
+        if (RegToAssign != AArch64::NoRegister)
+          Hints.push_back(RegToAssign);
```
This code is a little convoluted. I think you could also avoid the extra nested loop in `GetRegStartingAt`, by doing the following:

If Reg is e.g. `Z1_Z5_z9_z13`, then loop from `Z0_Z4_Z8_Z12`..`Z3_Z7_Z11_Z15` and check if any of them has allocated a phys reg. If not, then you can add `Z1_Z5_Z9_Z13`.

```
SmallVector<MCPhysReg> Regs;
unsigned FirstReg = Reg - OpIdx + 1;
for (unsigned I = 0; I < LdOps; ++I)
  Regs.push_back(FirstReg + I);

if (all_of(Regs,
           [&](MCPhysReg R) { return !Matrix->isPhysRegUsed(R); }))
  Hints.push_back(FirstReg + OpIdx - 1);
```

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -1109,24 +1110,93 @@ bool AArch64RegisterInfo::getRegAllocationHints(
-        RegID == AArch64::ZPR2StridedOrContiguousRegClassID
-            ? &AArch64::ZPR2StridedRegClass
-            : &AArch64::ZPR4StridedRegClass;
+  for (const MachineInstr &Use : MRI.use_nodbg_instructions(VirtReg)) {
```
The code here seems to be looping through uses of VirtReg, but it does that regardless of it's register class or feature flags.

What about returning early from the function if the function has no SME or is not in streaming-mode?
And also moving this loop inside a `if (RegID == .. || RegID == ...)` condition?

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -1109,24 +1110,93 @@ bool AArch64RegisterInfo::getRegAllocationHints(
+
+        // If the FORM_TRANSPOSE nodes use the ZPRMul classes, the starting
+        // register of the first load should be a multiple of 2 or 4.
+        if (IsMulZPR && (FirstReg - AArch64::Z0) % LdOps != 0)
```
Could you move this filtering to where it populates the `StridedOrder` vector?

sdesmalen-arm commented on llvm/test/CodeGen/AArch64/sme2-multivec-regalloc.mir:
```diff
@@ -0,0 +1,183 @@
+
+...
+---
+name:            form_4x_tuple_many_live
```
Thanks for cleaning these up!

sdesmalen-arm commented on llvm/test/CodeGen/AArch64/sme2-multivec-regalloc.mir:
```diff
@@ -0,0 +1,149 @@
+---
+name:            form_4x_tuple_many_live
+tracksRegLiveness: true
+liveins:
+  - { reg: '$x0', virtual-reg: '%0' }
+  - { reg: '$x1', virtual-reg: '%1' }
+  - { reg: '$z0', virtual-reg: '%2' }
+  - { reg: '$z17', virtual-reg: '%3' }
```
nit: I think you can remove these as well, because this information will be otherwise recomputed.

kmclaughlin-arm committed:
- Rewrote search for operand which already has a register assigned
- Add early exit if the function does not have SME or is not in streaming mode
- Move loop over uses of VirtReg into if block checking the RegID
- Remove GetRegStartingAt

kmclaughlin-arm committed:
- Allow mix of x2 & x4 multivector loads and intrinsics

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -1108,25 +1114,82 @@ bool AArch64RegisterInfo::getRegAllocationHints(
+          if (IsMulZPR && (FirstSubReg - AArch64::Z0) % LdOps != 0)
+            continue;
+
+          for (unsigned Op = 0; Op < LdOps; ++Op) {
+            if (!is_contained(StridedOrder, FirstStridedReg + Op) ||
+                getSubReg(FirstStridedReg + Op, AArch64::zsub0) !=
+                    FirstSubReg + Op)
+              break;
+            Regs.push_back(FirstStridedReg + Op);
+          }
+
+          if (Regs.size() == LdOps && all_of(Regs, [&](MCPhysReg R) {
+                return !Matrix->isPhysRegUsed(R);
+              }))
```
nit: This could be rewritten in such a way that it doesn't need `SmallVector<MCPhysReg> Regs` as an intermediate step, e.g.
```
auto IsFreeConsecutiveRegs = [&](unsigned I) {
  // conditions
};
if (all_of(iota_range<unsigned>(0U, UseOps, /*Inclusive=*/false),
           IsFreeConsecutiveReg))
  Hints.push_back(Reg);
```

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -1108,25 +1114,82 @@ bool AArch64RegisterInfo::getRegAllocationHints(
+            return VRM->hasPhys(Op.getReg());
+          });
+
+      if (AssignedRegOp == Use.operands_end()) {
```
I think an example would be useful here.

kmclaughlin-arm committed:
- Renamed LdOps to UseOps
- Rewrote the case where no registers are already assigned to avoid creating
  the Regs vector.
- Added more comments with an example

kmclaughlin-arm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -1108,25 +1114,82 @@ bool AArch64RegisterInfo::getRegAllocationHints(
+          Use.getOpcode() != AArch64::FORM_TRANSPOSED_REG_TUPLE_X4_PSEUDO)
+        continue;
+
+      unsigned LdOps = Use.getNumOperands() - 1;
```
In the last commit I did change `StridedRC` to be based on `RegID`; this was to support the case where the sizes of the multi-vectors used by the loads and the pseudo are different as you've described here. I've now also renamed `LdOps` to `UseOps`.

kmclaughlin-arm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -1108,25 +1114,82 @@ bool AArch64RegisterInfo::getRegAllocationHints(
+        for (unsigned I = 0; I < StridedOrder.size(); ++I) {
+          MCPhysReg Reg = StridedOrder[I];
+          SmallVector<MCPhysReg> Regs;
+          unsigned FirstStridedReg = Reg - OpIdx + 1;
```
Thanks, I added the `is_contained` later to make sure `getSubReg` would not fail but missed that it could also fail here.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -1099,6 +1100,11 @@ bool AArch64RegisterInfo::getRegAllocationHints(
     const VirtRegMap *VRM, const LiveRegMatrix *Matrix) const {
   const MachineRegisterInfo &MRI = MF.getRegInfo();
```
nit: this can be moved down (after the if(..) condition)

sdesmalen-arm reviewed (approved):
LGTM with nit addressed.

Would it make sense to update the title/commit message to reflect that you've also made it less strict? i.e. extracting one column from 4 x 2-vector loads can be used with an instruction that requires sequential regs now.

kmclaughlin-arm committed:
- Moved MRI def after if condition for hasSME()

kmclaughlin-arm commented:
> Would it make sense to update the title/commit message to reflect that you've also made it less strict? i.e. extracting one column from 4 x 2-vector loads can be used with an instruction that requires sequential regs now.

Thanks for approving this! I've reworded the title a bit and changed the commit message to include allowing x2 & x4 multivector loads and intrinsics.
